proxystore/store/dim/websockets.py

```diff
@@ -0,0 +1,355 @@
+from proxystore.serialize import deserialize
+from proxystore.serialize import SerializationError
+from proxystore.serialize import serialize
+from proxystore.store.base import Store
+from proxystore.store.dim.utils import get_ip_address
+from proxystore.store.dim.utils import Status
+
+MAX_CHUNK_LENGTH = 64 * 1024
+MAX_SIZE_DEFAULT = 1024**3
+
+logger = logging.getLogger(__name__)
+server_process = None
+
+
+class WebsocketStoreKey(NamedTuple):
+    """Key to objects in a WebsocketStore."""
+
+    websocket_key: str
+    obj_size: int
+    peer: str
+
+
+class WebsocketStore(Store[WebsocketStoreKey]):
+    """Distributed in-memory store using websockets."""
+
+    addr: str
+    provider_id: int
+    max_size: int
+    chunk_size: int
+    _loop: asyncio.events.AbstractEventLoop
+
+    def __init__(
+        self,
+        name: str,
+        *,
+        interface: str,
+        port: int,
+        max_size: int = MAX_SIZE_DEFAULT,
+        cache_size: int = 16,
+        stats: bool = False,
+    ) -> None:
+        """Initialization of a Websocket client.
+
+        This client will initialize a local Websocket
+        server (Peer service) that it will store data to.
+
+        Args:
+            name (str): name of the store instance.
+            interface (str): the network interface to use
+            port (int): the desired port for communication
+            max_size (int): the maximum size to be
+                communicated via websockets (default: 1G)
+            cache_size (int): size of LRU cache (in # of objects). If 0,
+                the cache is disabled. The cache is local to the Python
+                process (default: 16).
+            stats (bool): collect stats on store operations (default: False).
+
+        """
+        global server_process
+
+        logger.debug('Instantiating client and server')
+
+        self.max_size = max_size
+        self.chunk_size = MAX_CHUNK_LENGTH
+
+        self.host = get_ip_address(interface)
+        self.port = port
+
+        self.addr = f'ws://{self.host}:{self.port}'
+
+        if server_process is None:
+            server_process = Process(target=self._start_server)
+            server_process.start()
+
+        self._loop = asyncio.new_event_loop()
+        self._loop.run_until_complete(self.server_started())
+
+        super().__init__(
+            name,
+            cache_size=cache_size,
+            stats=stats,
+            kwargs={
+                'interface': interface,
+                'port': self.port,
+                'max_size': self.max_size,
+            },
+        )
+
+    def _start_server(self) -> None:
+        """Launch the local Websocket server process."""
+        logger.info(
+            f'starting server on host {self.host} with port {self.port}',
+        )
+
+        # create server
+        ps = WebsocketServer(
+            self.host,
+            self.port,
+            self.max_size,
+        )
+        asyncio.run(ps.launch())
```

Comment:
This is creating a new event loop different from `self._loop`. Maybe we should use `self._loop.create_task()`?

Reply:
I kept it separate because it's in a separate process. I didn't have the time to determine if there were any performance penalties to this and what it even means for processes to share event loops. Will update though.

Reply:
yeah, ok, this intuitively doesn't work because processes don't share memory.

Reply:
Oh yeah! You are right. I will mark resolved.